### PR TITLE
feat(provider): OP engine api trait ext + superchain signal type

### PR DESCRIPTION
**Pull request by merklefruit**
## Motivation

Adds a `provider` crate (`op-alloy-provider`) with some extension API traits for the OP engine API.
Additionally, adds the `SuperchainSignal` type along with the canonical decoding impl for the V0 type.

Ref: https://github.com/paradigmxyz/op-rs/issues/11

### CODE REVIEW

File: crates/rpc-types-engine/src/superchain.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+/// protocol version. The version must be encoded as 32 bytes of DATA in JSON RPC usage.
+///
+/// See also: <https://specs.optimism.io/protocol/superchain-upgrades.html#protocol-version>
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum ProtocolVersion {
+    /// Version-type 0.
+    V0(ProtocolVersionFormatV0),
+}
+
+cfg_if! {
```
Comment (mattsse) [inline crates/rpc-types-engine/src/superchain.rs:45]: I don't think this is required

File: crates/rpc-types-engine/src/superchain.rs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct ProtocolVersionFormatV0 {
+    pub build: B64,
+    pub major: u32,
+    pub minor: u32,
+    pub patch: u32,
+    pub pre_release: u32,
+}
+
+impl From<ProtocolVersionFormatV0> for [u8; 31] {
```
Comment (mattsse) [inline crates/rpc-types-engine/src/superchain.rs:101]: let's impl this as function, from conversions like this are just obfuscating things.

File: crates/rpc-types-engine/src/superchain.rs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        // <minor> ::= <big-endian uint32>
+        // <patch> ::= <big-endian uint32>
+        // <pre-release> ::= <big-endian uint32>
+
+        Ok(Self {
+            build: B64::from_slice(&value[7..15]),
+            major: u32::from_be_bytes(value[15..19].try_into()?),
+            minor: u32::from_be_bytes(value[19..23].try_into()?),
+            patch: u32::from_be_bytes(value[23..27].try_into()?),
+            pre_release: u32::from_be_bytes(value[27..31].try_into()?),
```
Comment (mattsse) [inline crates/rpc-types-engine/src/superchain.rs:143]: this will panic if called with a slice < len32

needs an additional len check

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/rpc-types-engine/Cargo.toml [hunk 4]
```diff
@@ -20,7 +20,7 @@ op-alloy-genesis.workspace = true
 op-alloy-consensus.workspace = true
 
 derive_more = { workspace = true, features = ["display"] }
-cfg-if.workspace = true
+thiserror.workspace = true
```
Comment (mattsse) [inline crates/rpc-types-engine/Cargo.toml:23]: if you want no-std then this won't work.

I'd suggest impl via derive_more

File: crates/rpc-types-engine/src/superchain.rs [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+pub enum ProtocolVersionError {
+    #[error("Unsupported version: {0}")]
+    UnsupportedVersion(u8),
+    #[error("Invalid version format length. Got {0}, expected 31")]
+    InvalidLength(usize),
+    #[error("Invalid version format encoding")]
+    FromSlice(#[from] TryFromSliceError),
+}
+
+impl From<ProtocolVersion> for B256 {
```
Comment (mattsse) [inline crates/rpc-types-engine/src/superchain.rs:55]: please convert these to functions, these from conversions are horrible to work with.

File: crates/rpc-types-engine/src/superchain.rs [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    /// ```text
+    /// <reserved><build><major><minor><patch><pre-release>
+    /// <reserved> ::= <7 zeroed bytes>
+    /// <build> ::= <8 bytes>
+    /// <major> ::= <big-endian uint32>
+    /// <minor> ::= <big-endian uint32>
+    /// <patch> ::= <big-endian uint32>
+    /// <pre-release> ::= <big-endian uint32>
+    /// ```
+    pub fn into_slice(self) -> [u8; 31] {
```
Comment (mattsse) [inline crates/rpc-types-engine/src/superchain.rs:145]: this is not a slice.

perhaps just encode?

**Review by mattsse — CHANGES_REQUESTED** [review]
last nits

File: crates/rpc-types-engine/src/superchain.rs (hunk 5 above)
Comment (merklefruit) [reply crates/rpc-types-engine/src/superchain.rs]: Got it, I thought the `[u8; 31]` was the issue

File: crates/rpc-types-engine/src/superchain.rs (hunk 6 above)
Comment (merklefruit) [reply crates/rpc-types-engine/src/superchain.rs]: renamed to `encode` and `decode` respectively

File: crates/rpc-types-engine/Cargo.toml (hunk 4 above)
Comment (merklefruit) [reply crates/rpc-types-engine/Cargo.toml]: Done with derive_more

File: crates/rpc-types-engine/src/superchain.rs [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    /// Identifies additional functionality in backwards compatible manner
+    pub minor: u32,
+    /// Identifies backward-compatible bug-fixes
+    pub patch: u32,
+    /// Identifies unstable versions that may not satisfy the above
+    pub pre_release: u32,
+}
+
+#[cfg(feature = "std")]
+impl std::fmt::Display for ProtocolVersionFormatV0 {
```
Comment (mattsse) [inline crates/rpc-types-engine/src/superchain.rs:180]: this can be core::fmt::Display

**Review by mattsse — APPROVED** [review]
final nit hehe

**Comment by merklefruit** [comment]
@mattsse done and added decoding tests from op-geth test vectors.
WASM target is failing for something related to "getrandom" lib - not sure how to address this

**Pull request merged**